### ISSUE TITLE
Feature: Support auto tls using Let's Encrypt

### ISSUE THREAD

**Issue by tamalsaha** [nats-io/nats-server#1893]
## Feature Request
Support auto tls using Let's Encrypt

#### Use Case:
As a developer, I want to deploy a TLS encrypted nats-sever publicly. My nats publishers are on the edge and streams data to the nats server. Today this is possible, but the certificate management is done separately. This is an extra operational complexity.

#### Proposed Change:
Integrate cert issue using acme protocol from Let's Encrypt. The code is pretty simple:
https://stackoverflow.com/a/40494806/244009

It could be an extra section in nats config:

```
tls: {
  acme: {
    hosts: ["example.com"]
    dir: /tmp/certs
    accetTOS: true
  }
}
```

#### Who Benefits From The Change(s)?
Any developer who wants to deploy nats with a public endpoint.

#### Alternative Approaches
Manage certificates externally. If you are on Kubernetes, you can use cert-manager.io . But I believe you still need to update the nats process to pick up the new certificate.

**Comment by alexellis** [comment]
I landed here looking for similar to this.

Can NATS be put behind a reverse proxy like Caddy or Nginx, with a HTTP01 challenge or similar?

**Comment by derekcollison** [comment]
@tamalsaha I use LE for the demo network as is, so maybe we can coordinate. I just run the LE process to update and signal the server to reload and all is good.

**Comment by derekcollison** [comment]
@alexellis You can put us behind something that does real layer 4, like HAProxy. I have done that myself. Anything above you maybe can but TLS from the normal NATS client protocol will probably not work, meaning you would only accept websocket connections.

May I ask why you would want to do that? For instance I would setup a layer of Leafnodes that connect back to a core NATS cluster as a type of DMZ.

**Comment by alexellis** [comment]
I want to run NATS with a valid TLS certificate for using JetStream, because it will be exposed on the Internet. Websockets would be fine.

**Comment by derekcollison** [comment]
You can do that today without anything in front. The demo system (demo.nats.io) has LE TLS, but allows non-tls optionally on same port, supports NATS client, websockets and MQTT, and has JetStream enabled. No reason to put anything in front per say IMO, sans DDOS vectors but I sue ufw for that..

**Comment by derekcollison** [comment]
Yes I use certbot to update and then simply reload the server, does not go done, use systemd.

**Comment by derekcollison** [comment]
Here is the complete demo server config. I run certbot and send a signal to the server to reload.

```
port: 4222
https: 8222
server_name: us-central-nats-demo

max_connections: 250000
max_subscriptions: 200000

reconnect_error_reports: 3600
max_traced_msg_len: 64


logfile_size_limit: 2GB
log_file: "/var/log/nats-server.log"

jetstream {
  store_dir: /var/jetstream
  max_mem_store:  10GiB
  max_file_store: 410GiB
}

tls {
  cert_file: "/etc/letsencrypt/live/demo.nats.io/fullchain.pem"
  key_file:  "/etc/letsencrypt/live/demo.nats.io/privkey.pem"
  timeout:    "5s"
}

# Allow both TLS and non-TLS to work on same port.
allow_non_tls: true

leafnodes {
  port: 7422
  tls {
    cert_file: "/etc/letsencrypt/live/demo.nats.io/fullchain.pem"
    key_file:  "/etc/letsencrypt/live/demo.nats.io/privkey.pem"
    timeout:    "5s"
  }
}

no_auth_user: demo-user

demo_perms = {
  publish = {
    # Do not allow deletion of MQTT streams
    deny = ["$JS.API.STREAM.DELETE.$MQTT_msgs", "$JS.API.STREAM.DELETE.$MQTT_rmsgs", "$JS.API.STREAM.DELETE.$MQTT_sess"]
  }
}

accounts {
  default: {
    jetstream: {
      max_mem:		8GiB
      max_store:	400GiB
      max_streams:	1024
      max_consumers:	8192
    }
    users = [ { user: demo-user, permissions: $demo_perms} ]
  }
  $SYS: {
    users = [ { nkey: UDEMO3ZANTMUGPSBS3H54WKJN3TNVGQBJUQFCT7H4MUQLCRRQ26CWIIP } ]
  }
}


websocket {
  port: 8443
  compression: true
  handshake_timeout: "5s"
  tls {
    cert_file: "/etc/letsencrypt/live/demo.nats.io/fullchain.pem"
    key_file:  "/etc/letsencrypt/live/demo.nats.io/privkey.pem"
    timeout:    "5s"
  }
  no_auth_user: demo-user
}

mqtt {
  port: 1883
  tls {
    cert_file: "/etc/letsencrypt/live/demo.nats.io/fullchain.pem"
    key_file:  "/etc/letsencrypt/live/demo.nats.io/privkey.pem"
    timeout:    "5s"
  }
  no_auth_user: demo-user
  ack_wait: "1m"
  max_ack_pending: 1024
}
```

**Comment by simar7** [comment]
Would be nice to have it built-in. I understand there are other ways to do it but operationally, there's still friction we can reduce with this being autotls.

**Comment by derekcollison** [comment]
How do you see autotls helping here? I am not too familiar so apologies if dumb question.

**Comment by charbonnierg** [comment]
I played a bit with this yesterday, and managed to start a NATS server with TLS enabled for:
- Standard Listener
- Monitoring Listener
- Websocket Listener
- MQTT Listener
- Leafnode Listener

Simply by using `certmagic` and overriding the `TLSConfig` in server options with a config obtained using `certmagic.NewDefault().TLSConfig()`.

I can share more code if some people are interested, but I wonder if there are any downside in overriding the TLSConfig property in various fields in server options ?
